### PR TITLE
Adding stronger password security

### PR DESCRIPTION
**Pull request by jloveland**
Adding stronger password requirements for improving security based on OWASP. Fixing #848

### CODE REVIEW

File: modules/users/client/directives/password-validator.client.directive.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+angular.module('users')
+  .directive('passwordValidator', ['$window', function($window) {
+    return {
+      require: 'ngModel',
+      link: function(scope, element, attrs, modelCtrl) {
+        modelCtrl.$parsers.unshift(function (password) {
+          var owaspPasswordStrengthTest = $window.owaspPasswordStrengthTest;
```
Comment (mleanos) [inline modules/users/client/directives/password-validator.client.directive.js]: Can we have a client service for `owasp`? Some user's might want to extend it, or have a bit more control over it's behavior.

**Comment by mleanos** [comment]
@jloveland This all looks pretty good. I left a line comment. I haven't taken this for a test drive yet, but I plan to sometime tonight. Overall, I like what I see. Great job!

**Comment by jloveland** [comment]
Ok, I will add a client service tonight.

**Comment by jloveland** [comment]
ok, added a client service, check it out.

**Comment by mleanos** [comment]
@jloveland I'll look at this again tonight. Got side-tracked yesterday with the jshint issue.

File: modules/users/tests/server/user.server.model.tests.js [hunk 2]
```diff
@@ -217,9 +216,87 @@ describe('User Model Unit Tests:', function () {
...
+
+      _user1.validate(function (err) {
+        err.errors.password.message.should.equal("The password must be fewer than 128 characters.");
+        done();
+      });
+    });
+
+    it('should not allow more than 3 or more repeating characters ', function (done) {
+      var _user1 = new User(user1);
+      _user1.password = 'L0veSexxxSecre+God';
```
Comment (mleanos) [inline modules/users/tests/server/user.server.model.tests.js]: might want to change that :)

**Comment by mleanos** [comment]
@jloveland I took this for a test drive, and everything is working perfectly. I see you added tests as well :) 

I left a line comment about one of the passwords you have set in one of the new tests. The only other suggestion might be to remove the pop-over when the validation is at 100%. This is merely a suggestion; I just wanted to point it out.

Everything else looks really good. I think this is good to go. Thank you for knocking this out. It's really a great enhancement.

File: modules/users/tests/server/user.server.model.tests.js (hunk 2 above)
Comment (jloveland) [inline modules/users/tests/server/user.server.model.tests.js]: haha..yeah, I copied the test passwords from the owasp lib [here](https://github.com/nowsecure/owasp-password-strength-test/blob/master/test.js#L30), I can change that to be less *interesting*..

**Comment by jloveland** [comment]
I can also disable the pop-over once the test passes.

**Comment by jloveland** [comment]
this is still WIP. I noticed that
1. I missed the reset password form.
2. password reset request fails because of `undefined` password
So, I am working on these now.

I need some advice on how to implement closing the popover once the password field is valid. I tried `popover-is-open="{{userForm.password.$error}}"` and that was causing conflicts..

**Comment by jloveland** [comment]
I added password validation and verification, fixed tests, etc. I did not implement disabling the popover, maybe that can be done on a different PR? 

Also, as you can see, I had some issues with squashing my commits properly, so I pushed anyway. Maybe someone can help me with the rebase?

**Comment by simison** [comment]
Great! Testing this out. 

~~UX wise it would be clearer if those requirements would be visible already from the beginning, not only when I hit "sign up". Now it gives blue 80% (=good?) strength for complicated but short passwords.~~

__Edit:__ Oh, there indeed is a popup telling all that. It just doesn't appear until I type something into the field and then click out and focus again to the input.

Incidentally I have an easy to use Angular module for this stuff. :-) https://github.com/trustroots/trustpass Check if you like it.

**Comment by jloveland** [comment]
I just squashed my commits, let me know what you think. @simison thanks for the suggestion. maybe we can evaluate your library and refactor with another PR once this is merged?

**Comment by simison** [comment]
@jloveland sure! I'd try to fix that modal not appearing on first focus anyway. I think I've had that before too when trying to add content dynamically into popups/tooltips. Make sure you have the latest bootstrap-ui though.

**Comment by mleanos** [comment]
@jloveland I've tested this out again, with the password recovery too. Looks good to me. Do you still need help with the rebasing?

To get my PR branches rebased, I usually get my local master branch current, with upstream (meanjs), and then checkout my PR branch, and perform `git rebase master` Sometimes, I'll get a warning that my origin & local have diverged. In which case, I just push to my origin with `--force`

**Comment by jloveland** [comment]
@mleanos I think I got the rebase worked out. please let me know if something looks wrong.

@simison I saw a lot of references to similar issues when researching the *dynamic* popover. I could solve the popover issue by reverting to *static* popover...placing the message in the [html](https://github.com/jloveland/mean/blob/owasp-password-strength/modules/users/client/views/authentication/signup.client.view.html#L37) like so:
`popover="Please enter a passphrase or password with greater than 10 characters, numbers, lowercase, uppercase, and special characters."`

instead of the dynamic approach:
`popover="{{popoverMsg}}"`

This is not ideal, but it would fix the problem. What do you think?

**Comment by mleanos** [comment]
@jloveland I'm not an expert with Git, so correct me if I'm wrong. I think having those other commits (other than from you) is the result of not squashing after your rebase. I usually try get my PR's down to one commit.

**Comment by jloveland** [comment]
@mleanos, you were correct, I needed to `git rebase master` then squash commits.

Are there any Angular.js experts out there that can advise me on how to solve the issue with the dynamic popover message not displaying on focus properly? Otherwise, my next change will be to revert to static popover defined in the html.

**Comment by simison** [comment]
It has something to do with dynamic content; try to make sure that variable always has some non-empty string content or try fiddling with `tooltip-enable="true|false"` or `tooltip-is-open="true|false"`.

**Comment by jloveland** [comment]
- fixed the dynamic popover message scope issues by moving the message variable the service.
- made popover disabled when password strength passes
- made button sizes consistent with the rest of the views

**Comment by mleanos** [comment]
@jloveland Very good. Glad you got those issues worked out. I've tested this, and everything is working wonderfully! Thanks. 

I think this is ready to go. @simison How's it look to you?

**Comment by codydaig** [comment]
LGTM 

Thanks @jloveland! :-D

**Comment by lirantal** [comment]
Generally reviewed it, LGTM.

**Comment by mleanos** [comment]
@lirantal I had the same concern over existing user's. I tested this, and these changes didn't have an impact on my existing users.

These changes merely enforce the new password requirements, when a user attempts to change their password; or when a new user signs up.

@jloveland ?

**Comment by jloveland** [comment]
@lirantal and @mleanos this does not affect existing users until they change their password. It will have immediate affect on users who signup.

**Comment by lirantal** [comment]
Cool, let's continue with this then.

File: modules/users/client/directives/password-verify.client.directive.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+angular.module('users')
+  .directive("passwordVerify", function() {
+    return {
+      require: "ngModel",
```
Comment (simison) [inline modules/users/client/directives/password-verify.client.directive.js]: @jloveland JSHint is complaining about not using single quotes here. Could you fix and send another PL?

File: modules/users/client/directives/password-verify.client.directive.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+          return combined;
+        }, function(value) {
+          if (value) {
+            modelCtrl.$parsers.unshift(function(viewValue) {
+              var origin = scope.passwordVerify;
+              if (origin !== viewValue) {
+                modelCtrl.$setValidity("passwordVerify", false);
+                return undefined;
+              } else {
+                modelCtrl.$setValidity("passwordVerify", true);
```
Comment (simison) [inline modules/users/client/directives/password-verify.client.directive.js]: should be single quotes

File: modules/users/client/services/password-validator.client.service.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  function ($window) {
+    var owaspPasswordStrengthTest = $window.owaspPasswordStrengthTest;
+
+    return {
+      getResult: function (password) {
+        var result = owaspPasswordStrengthTest.test(password);
+        return result;
+      },
+      getPopoverMsg: function () {
+        var popoverMsg = "Please enter a passphrase or password with greater than 10 characters, numbers, lowercase, upppercase, and special characters.";
```
Comment (simison) [inline modules/users/client/services/password-validator.client.service.js]: should be single quotes

Comment (jloveland) [inline modules/users/client/services/password-validator.client.service.js]: Thanks @simison , I didn't see this show after running `grunt jshint` on that branch. Output from the command:
```
grunt jshint
Running "jshint:all" (jshint) task
>> 105 files lint free.

Done, without errors.
```
Any thoughts on why I don't get the errors?
I can submit another PR to correct this though...

Comment (simison) [inline modules/users/client/services/password-validator.client.service.js]: A good question. I just copied the directive to my meanjs-based project and got this. Is /client/services/ a new folder?

BTW Thanks for writing these!

File: modules/users/client/directives/password-verify.client.directive.js (hunk 3 above)
Comment (ilanbiala) [inline modules/users/client/directives/password-verify.client.directive.js]: also in the directive name above?

**Pull request merged**